### PR TITLE
9.4.x: Only release connection when channel got released to avoid double connection release race condition

### LINKED ISSUE #6285: HTTP2 client: IllegalStateException: Cannot release an already released entry
**Jetty version**
9.4.x; maybe 10.0.x and 11.0.x too

**Description**
There seems to be a race condition in the HTTP2 client that makes it very rarely fail with a `IllegalStateException` when multiplexing is enabled:

```
java.util.concurrent.CompletionException: java.lang.IllegalStateException: Cannot release an already released entry
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:783)
	at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:792)
	at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2153)
	at org.mortbay.jetty.load.generator.LoadGenerator.process(LoadGenerator.java:358)
	at org.mortbay.jetty.load.generator.LoadGenerator.lambda$begin$0(LoadGenerator.java:187)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: Cannot release an already released entry
	at org.eclipse.jetty.util.Pool$Entry.tryRelease(Pool.java:623)
	at org.eclipse.jetty.util.Pool.release(Pool.java:414)
	at org.eclipse.jetty.client.AbstractConnectionPool.deactivate(AbstractConnectionPool.java:378)
	at org.eclipse.jetty.client.AbstractConnectionPool.release(AbstractConnectionPool.java:364)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:352)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:336)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:315)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:309)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:286)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:263)
	at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:600)
	at org.eclipse.jetty.client.HttpRequest.sendAsync(HttpRequest.java:778)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:765)
	at org.mortbay.jetty.load.generator.LoadGenerator$Sender.send(LoadGenerator.java:641)
	at org.mortbay.jetty.load.generator.LoadGenerator$Sender.send(LoadGenerator.java:601)
	at org.mortbay.jetty.load.generator.LoadGenerator$Sender.access$500(LoadGenerator.java:562)
	at org.mortbay.jetty.load.generator.LoadGenerator.sendResourceTree(LoadGenerator.java:449)
	at org.mortbay.jetty.load.generator.LoadGenerator.process(LoadGenerator.java:334)
	... 5 more
```

This was first seen in `jetty-load-generator`'s `HTTP2WebsiteLoadGeneratorTest.testHTTP2WithPush()` test but it seems generic, and just hard to trigger.

### PR DESCRIPTION
**Pull request by lorban**
Signed-off-by: Ludovic Orban <lorban@bitronix.be>

### CODE REVIEW

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 1]
```diff
@@ -461,7 +461,10 @@ public void close()
         // iterate the copy and close its entries
         for (Entry entry : copy)
         {
-            if (entry.tryRemove() && entry.pooled instanceof Closeable)
+            boolean removed = entry.tryRemove();
+            if (!removed)
+                LOGGER.warn("Pooled object still in use: {}", entry);
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Maybe you want to stay at DEBUG here.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 1 above)
Comment (lorban) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Yes, that sounds reasonable.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 2]
```diff
@@ -461,7 +461,13 @@ public void close()
...
         for (Entry entry : copy)
         {
-            if (entry.tryRemove() && entry.pooled instanceof Closeable)
+            boolean removed = entry.tryRemove();
+            if (!removed)
+            {
+                if (LOGGER.isDebugEnabled())
+                    LOGGER.debug("Pooled object still in use: {}", entry);
+            }
+            if (removed && entry.pooled instanceof Closeable)
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Can you flip the `if/else` for readability?

```java
boolean removed = entry.tryRemove();
if (removed) {
  if (instanceof Closeable) 
    ...
} else {
  if (LOGGER.isDebugEnabled())
     ...
}

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 2 above)
Comment (lorban) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Done.

**Review by sbordet — APPROVED** [review]

**Pull request merged**